### ISSUE TITLE
Adding support for one more String function: strtok()

### ISSUE THREAD

**Issue by zarif98** [bpftrace/bpftrace#1886]
It would be nice to have strtok() functionality to parse strings and/or compare strings from my preloaded table with the kernel. The limitation right now makes it almost impossible for certain tasks to be executed.

Handwriting the solution does not seem to work since we end up with unknown opcode errors regardless of any variation

**Comment by danobi** [comment]
Could you share some more concrete examples of what tasks you're trying to accomplish?

**Comment by zarif98** [comment]
I am testing some hacked together drives that I made with an ARM processor. Without going into too much detail, I would like to parse some input from the user that comes in either a CSV format, words separated by a space, or something with a delimiter so I can do my own computation on it. That is my goal as of now.

**Comment by danobi** [comment]
String processing in a BPF prog is tricky. It might require new kernel helpers. It'll definitely require @Birch-san 's map-backed strings. Without map-backed strings you can process at most ~150 characters. And then we'll have to bound the loop too.

This is one of those features that sounds simple but is actually quite tricky due to how restrictive BPF prog execution is.

**Comment by zarif98** [comment]
Yes, I have been trying to try and see if I could come up with something using restrictive C but it is literally impossible with my skillset. I have been passing the strings to userspace when parsing but I do lose some of the performance benefit that I could have seen when running within BPF. I wonder if anyone could divulge more information about some kernel helpers. Maybe I can help write? Never wrote any kernel functions before but I could give it a shot.

**Comment by danobi** [comment]
Here's the `strtol()` helper:

https://github.com/torvalds/linux/blob/007b350a58754a93ca9fe50c498cc27780171153/include/uapi/linux/bpf.h#L3587
https://github.com/torvalds/linux/blob/df04fbe8680bfe07f3d7487eccff9f768bb02533/kernel/bpf/helpers.c#L500-L513

**Comment by zarif98** [comment]
Hi! Thank you for the link but I am not sure how this would be useful for what I am trying to accomplish?

**Comment by danobi** [comment]
BPF helpers are implemented in the kernel. It's unlikely we'll be able to implement strtok() without help from the kernel.

**Comment by zarif98** [comment]
So, strsep() is implemented within the kernel. Do I

**Comment by amscanne** [comment]
Closing as stale. Can be reopened if there's renewed discussion.
